src/dos/dos_code_page.cpp

```diff
@@ -0,0 +1,2223 @@
+/*
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *
+ *  Copyright (C) 2025-2025  The DOSBox Staging Team
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "dos_code_page.h"
+
+#include "../ints/int10.h"
+#include "bios.h"
+#include "checks.h"
+#include "dos_locale.h"
+#include "math_utils.h"
+#include "mem.h"
+#include "string_utils.h"
+#include "support.h"
+#include "unicode.h"
+
+#include <algorithm>
+#include <array>
+#include <fstream>
+#include <optional>
+#include <set>
+#include <unordered_map>
+#include <vector>
+
+CHECK_NARROWING();
+
+// References:
+// - https://www.seasip.info/DOS/CPI/cpi.html
+// - http://kbd-project.org/docs/font-formats/font-formats-3.html
+
+// ***************************************************************************
+// Constants
+// ***************************************************************************
+
+static const std::string ResourceDir = "freedos-cpi";
+
+enum class FileFormat {
+	// Not yet detected or unrecognized
+	Unknown,
+	// Used by MS-DOS and most clones
+	MsDos,
+	// Used by Windows NT, slight variation of MS-DOS format
+	WinNt,
+	// Used by DR-DOS, encodes screen fonts in much more efficient way
+	DrDos
+};
+
+// Supported file size limit, just for some extra safety.
+// - Almost all the CPI files in the wild are below 64 KB size (MS DOS does not
+//   support anything bigger)
+// - 'DISPLAY.CPI' file from PTS-DOS 2000 is below 88 KB
+// - 'EGA.CPI' file from Windows 2000 (English) is slightly below 128 KB
+constexpr uint32_t MaxFileSize = 1024 * 1024;
```

Comment:
`MaxFileSizeBytes` would be better (always include the unit).

Reply:
OK, changed.